pyro/poutine/reparam_messenger.py

```diff
@@ -0,0 +1,72 @@
+from collections import OrderedDict
+
+import torch
+
+from pyro.distributions.delta import Delta
+
+from .messenger import Messenger
+from .runtime import apply_stack
+
+
+class ReparamMessenger(Messenger):
+    """
+    Reparametrizes each affected sample site into one or more auxiliary sample
+    sites followed by a deterministic transformation [1].
+
+    To specify :class:`~pyro.distributions.reparameterize.Reparameterizer` s,
+    either pass a ``config`` dict to the constructor, configure
+    ``site["infer"]["reparam"] = my_reparameterizer`` for each desired sample
+    site, or use :func:`~pyro.poutine.infer_config` .
+
+    See `available reparameterizers <distributions.html#reparameterizers>`_
+
+    .. warning:: Reparameterizers are recursive; take care to avoid infinite
+        loops in your ``config`` filters.
+
+    [1] Maria I. Gorinova, Dave Moore, Matthew D. Hoffman (2019)
+        "Automatic Reparameterisation of Probabilistic Programs"
+        https://arxiv.org/pdf/1906.03028.pdf
+
+    :param dict config: Optional configuration mapping site name to
+        :class:`~pyro.distributions.reparameterize.Reparameterizer` object.
+    """
+    def __init__(self, config=None):
+        super().__init__()
+        if config is None:
+            config = {}
+        self.config = config
+
+    def _pyro_sample(self, msg):
```

Comment:
@eb8680 can you take a look to ensure I've implemented this correctly? I've followed `poutine.do`.

Reply:
I'll need to read the paper to understand so probably won't be able to get to this today

Reply:
Oh I'm not so concerned with matching up with the paper as ensuring that I have correctly transformed a single `sample` site into multiple `sample` sites plus a `pyro.deterministic` site.